File: src/dashboard/dashboard.controller.ts
import { Controller, Get } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Order } from '../orders/order.entity';
import { Product } from '../products/product.entity';

@Controller('dashboard')
export class DashboardController {
  constructor(
    @InjectRepository(Order) private readonly ordersRepo: Repository<Order>,
    @InjectRepository(Product) private readonly productsRepo: Repository<Product>,
  ) {}

  @Get()
  async metrics() {
    const orders = await this.ordersRepo.find({ relations: ['items', 'items.product'] });
    const revenue = orders.reduce((sum, o) => sum + o.items.reduce((s, i) => s + Number(i.product?.price || 0) * i.quantity, 0), 0);

    const salesReturn = 30000; // placeholder for design symmetry
    const purchase = 30000; // placeholder
    const income = 30000; // placeholder

    const topSelling: Record<string, number> = {};
    for (const o of orders) {
      for (const i of o.items) {
        if (!i.product) continue;
        topSelling[i.product.name] = (topSelling[i.product.name] || 0) + i.quantity;
      }
    }
    const topSellingArray = Object.entries(topSelling)
      .map(([name, qty]) => ({ name, qty }))
      .sort((a, b) => b.qty - a.qty)
      .slice(0, 5);

    const stockAlert = (await this.productsRepo.find()).filter((p) => p.stock <= 5);

    return {
      cards: { revenue, salesReturn, purchase, income },
      topSelling: topSellingArray,
      stockAlert,
    };
  }
}

File: java-api/src/main/java/com/integrador/api/dashboard/DashboardController.java
package com.integrador.api.dashboard;

import com.integrador.api.orders.Order;
import com.integrador.api.orders.OrderRepository;
import com.integrador.api.products.Product;
import com.integrador.api.products.ProductRepository;
import com.integrador.api.returns.ReturnRecordRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@RestController
@RequestMapping("/dashboard")
public class DashboardController {
    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final ReturnRecordRepository returnRecordRepository;

    public DashboardController(OrderRepository orderRepository, ProductRepository productRepository, ReturnRecordRepository returnRecordRepository) {
        this.orderRepository = orderRepository;
        this.productRepository = productRepository;
        this.returnRecordRepository = returnRecordRepository;
    }

    @GetMapping
    public ResponseEntity<?> metrics(@RequestParam(required = false) String period) {
        List<Order> orders = orderRepository.findAll().stream()
                .filter(o -> inPeriod(o.getCreatedAt(), period))
                .filter(o -> !"returned".equalsIgnoreCase(o.getStatus()))
                .toList();

        BigDecimal revenue = orders.stream()
                .flatMap(o -> o.getItems().stream())
                .map(i -> i.getProduct().getPrice().multiply(BigDecimal.valueOf(i.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal salesReturn = returnRecordRepository.findAll().stream()
                .filter(r -> inPeriod(r.getDate(), period))
                .map(r -> r.getValue() == null ? BigDecimal.ZERO : r.getValue())
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        BigDecimal purchase = BigDecimal.ZERO;
        BigDecimal income = revenue.subtract(purchase);

        Map<String, Integer> topSelling = new HashMap<>();
        for (Order order : orders) {
            order.getItems().forEach(item -> {
                if (item.getProduct() != null) {
                    topSelling.merge(item.getProduct().getName(), item.getQuantity(), Integer::sum);
                }
            });
        }

        List<Map<String, Object>> topSellingArray = topSelling.entrySet().stream()
                .sorted((a, b) -> b.getValue().compareTo(a.getValue()))
                .limit(5)
                .<Map<String, Object>>map(entry -> {
                    Map<String, Object> row = new HashMap<>();
                    row.put("name", entry.getKey());
                    row.put("qty", entry.getValue());
                    return row;
                })
                .toList();

        List<Product> stockAlert = productRepository.findAll().stream().filter(p -> p.getStock() <= 5).toList();

        return ResponseEntity.ok(Map.of(
                "cards", Map.of("revenue", revenue, "salesReturn", salesReturn, "purchase", purchase, "income", income),
                "topSelling", topSellingArray,
                "stockAlert", stockAlert
        ));
    }

    private boolean inPeriod(LocalDateTime value, String period) {
        if (value == null || period == null || period.isBlank() || "all".equalsIgnoreCase(period)) return true;
        LocalDate today = LocalDate.now();
        return switch (period.toLowerCase()) {
            case "today" -> value.toLocalDate().isEqual(today);
            case "7d" -> !value.toLocalDate().isBefore(today.minusDays(6));
            case "30d" -> !value.toLocalDate().isBefore(today.minusDays(29));
            default -> true;
        };
    }
}
